### PR TITLE
(2782) Fix invalid comment update and other edge cases

### PR DESCRIPTION
**Pull request by yndajas**
## Changes in this PR

Previously:
- If when editing a comment, you submit the form with no body, an error would show up due to a missing instance variable
- Tests were missing for this in both the `update` and `create` controller actions (but the latter didn't cause an error)
- When an activity had no title, the heading when adding/editing comments was confusing: "Add a comment about"/"Edit comment about"
- Some comment tests were checking for the absence of the wrong translation
- If a create/update failed, the breadcrumbs would be lost when re-rendering the form

This fixes all of the above

### Before

image

### After

image

## Next steps

- [ ] Is an ADR required? An ADR should be added if this PR introduces a change to the architecture.
- [x] Is a changelog entry required? An entry should always be made in `CHANGELOG.md`, unless this PR is a small tweak which has no impact outside the development team.
- [ ] Do any environment variables need amending or adding?
- [ ] Have any changes to the XML been checked with the IATI validator? See [XML Validation](https://github.com/UKGovernmentBEIS/beis-report-official-development-assistance/blob/develop/doc/xml-validation.md)

### CODE REVIEW

File: app/controllers/activity_comments_controller.rb [hunk 1]
```diff
@@ -93,4 +92,18 @@ def report_id
   def comment_params
     params.require(:comment).permit(:body, :report_id)
   end
+
+  def prepare_trail_and_breadcrumb(is_new:)
```
Comment (Gweaton) [inline app/controllers/activity_comments_controller.rb:96]: I'm finding this a little hard to parse - I wonder if it might be a bit clearer to opt for duplication here, and have a `prepare_new_trail_and_breadcrumb` and a `prepare_edit_trail_and_breadcrumb` method?

Also, are there already tests for this behaviour?

**Review by Gweaton — COMMENTED** [review]
This looks good - just one thing re: that long method.

Also, let's add this to the CHANGELOG.

**Review by Gweaton — COMMENTED** [review]
This looks good - just one thing re: that long method.

Also, let's add this to the CHANGELOG.

File: app/controllers/activity_comments_controller.rb (hunk 1 above)
Comment (yndajas) [reply app/controllers/activity_comments_controller.rb]: I've split that method in two - looks better that way

I think the controller specs don't usually go into enough detail to cover trails/breadcrumbs. Can't see any existing tests in the specs I've touched referencing trails or breadcrumbs. Do you think we should add them, or leave the controller specs light?

Comment (Gweaton) [reply app/controllers/activity_comments_controller.rb]: It looks like we've tested breadcrumbs in feature specs in the past e.g. [here](https://github.com/UKGovernmentBEIS/beis-report-official-development-assistance/blob/a74265fed7afd7ae1341aa05019d3e0c89d279dd/spec/features/users_can_search_for_activities_spec.rb#L31). It might be worth adding specs like that in the new/edit comments feature spec, if that's not too much of a pain, so at least we can catch things if they go missing.

**Review by Gweaton — APPROVED** [review]
Sorry, meant to approve this but looks like I commented...

**Review by Gweaton — APPROVED** [review]
Nice one, thanks for adding those tests.
